lib/charms/observability_libs/v1/kubernetes_compute_resources_patch.py

```diff
@@ -0,0 +1,631 @@
+        # in those events firing without a workload.
+        self.framework.observe(charm.on.config_changed, self._on_config_changed)
+
+        if not refresh_event:
+            refresh_event = []
+        elif not isinstance(refresh_event, list):
+            refresh_event = [refresh_event]
+        for ev in refresh_event:
+            self.framework.observe(ev, self._on_config_changed)
+
+    def _on_config_changed(self, _):
+        self._patch()
+
+    def _patch(self) -> None:
+        """Patch the Kubernetes resources created by Juju to limit cpu or mem.
+
+        This method will keep on retrying to patch the kubernetes resource for a default duration of 3 minutes,
+        if the patching failure is due to a recoverable error (e.g: Network Latency).
+        This will put the charm in a `Maintenance` state until the patch is either successful or all retries have been consumed,
+        which in that case, will raise the same exception as the original exception that caused it and will put the charm in error state.
+        If the failure is not recoverable (e.g: wrong spec provided), an exception same as the original exception that caused it will be raised and will put the charm in error state.
+        """
+        try:
+            resource_reqs = self.resource_reqs_func()
+            limits = resource_reqs.limits
+            requests = resource_reqs.requests
+        except ValueError as e:
+            msg = f"Failed obtaining resource limit spec: {e}"
+            logger.error(msg)
+            self.on.patch_failed.emit(message=msg)
+            raise ValueError(msg) from e
+
+        for spec in (limits, requests):
+            if not is_valid_spec(spec):
+                msg = f"Invalid resource limit spec: {spec}"
+                logger.error(msg)
+                self.on.patch_failed.emit(message=msg)
+                raise RuntimeError(msg)
+
+        resource_reqs = ResourceRequirements(
+            limits=sanitize_resource_spec_dict(limits),  # type: ignore[arg-type]
+            requests=sanitize_resource_spec_dict(requests),  # type: ignore[arg-type]
+        )
+
+        try:
+            for attempt in tenacity.Retrying(
+                retry=self.PATCH_RETRY_IF,
+                stop=self.PATCH_RETRY_STOP,
+                wait=self.PATCH_RETRY_WAIT,
+                # if you don't succeed raise the last caught exception when you're done
+                reraise=True,
+            ):
+                with attempt:
+                    self._charm.unit.status = MaintenanceStatus(
+                        f"retrying patching resource limit... (attempt #{attempt.retry_state.attempt_number})"
+                    )
+                    self.patcher.apply(resource_reqs)
+
+        except exceptions.ConfigError as e:
+            msg = f"Error creating k8s client: {e}"
+            logger.error(msg)
+            self.on.patch_failed.emit(message=msg)
+            raise exceptions.ConfigError(msg) from e
```

Comment:
a bit confused here by why we're reraising the same exception type but with a different message. If the user sees `lightkube.exceptions.ConfigError`, they'd  rightfully assume that the exception originates from lightkube while in fact you're firing it. 
I'd be in favour of doing a `raise PatchFailedError(msg) from e` instead.

Reply:
or simply reraise it without changing the message. 
Also FYI you can:
`logger.exception(msg)` which will automatically attach the traceback to the error log.

Reply:
raised a custom `PatchFailedError` exception